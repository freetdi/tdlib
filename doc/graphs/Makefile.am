

# are these really free?
EXTRA_DIST = \
	contiki \
	coremark \
	dhrystone \
	dimacs \
	fuzix \
	maxsat_small \
	random_graphs \
	random_k_tree \
	stdlib \
	whetstone \
	zoo

# TODO: clean dist directories here.
dist-hook:
	:
