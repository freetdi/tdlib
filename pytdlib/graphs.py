from RandomGNM_250_1000 import *
from Graph import Graph
from Decomp import Decomp

#some graphs
V_P6 = [0,1,2,3,4,5]
E_P6 = [(0,1),(1,2),(2,3),(3,4),(4,5)]

V_K5 = [0,1,2,3,4]
E_K5 = [(0,1),(0,2),(0,3),(0,4),(1,2),(1,3),(1,4),(2,3),(2,4),(3,4)]

V_Petersen = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
E_Petersen = [(0, 1), (0, 4), (0, 5), (1, 2), (1, 6), (2, 3), (2, 7), \
              (3, 4), (3, 8), (4, 9), (5, 7), (5, 8), (6, 8), (6, 9), (7, 9)]

V_Petersen_double = [0,1,2,3,4,5,6,7,8,9,20,21,22,23,24,25,26,27,28,29]
E_Petersen_double = [(0, 1), (0, 4), (0, 5), (1, 2), (1, 6), (2, 3), (2, 7), \
              (3, 4), (3, 8), (4, 9), (5, 7), (5, 8), (6, 8), (6, 9), (7, 9), \
              (20,21),(20,24),(20,25),(21,22),(21,26),(22,23),(22,27), \
              (23,24),(23,28),(24,29),(25,27),(25,28),(26,28),(26,29),(27,29)]

V_Wagner = [0, 1, 2, 3, 4, 5, 6, 7]
E_Wagner = [(0, 1),(0, 4), (0, 7), (1, 2), (1, 5), (2, 3), (2, 6), (3, 4), \
            (3, 7), (4, 5), (5, 6), (6, 7)]

V_Pappus = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17]
E_Pappus = [(0, 1), (0, 5), (0, 6), (1, 2), (1, 7), (2, 3), (2, 8), (3, 4), \
            (3, 9), (4, 5), (4, 10), (5, 11), (6, 13), (6, 17), (7, 12), \
            (7, 14), (8, 13), (8, 15), (9, 14), (9, 16), (10, 15), (10, 17), \
            (11, 12), (11, 16), (12, 15), (13, 16), (14, 17)]

#grids
V_Grid_5_5 = range(25)
E_Grid_5_5 = [(1, 0), (2, 1), (3, 2), (4, 3), (5, 0), (6, 1), (6, 5), (7, 2), \
              (7, 6), (8, 3), (8, 7), (9, 4), (9, 8), (10, 5), (11, 6), \
              (11, 10), (12, 7), (12, 11), (13, 8), (13, 12), (14, 9), (14, 13), \
              (15, 10), (16, 11), (16, 15), (17, 12), (17, 16), (18, 13), (18, 17), \
              (19, 14), (19, 18), (20, 15), (21, 16), (21, 20), (22, 17), (22, 21), \
              (23, 18), (23, 22), (24, 19), (24, 23)]


#corner cases

V_Empty = []
E_Empty = []

V_OneNode = [1]
E_OneNode = []

V_Edgeless = [1,2,3,4,5]
E_Edgeless = []

V_Clique = [1,2,3,4,5]
E_Clique = [(1,2),(1,3),(1,4),(1,5),(2,3),(2,4),(2,5),(3,4),(3,5),(4,5)]

V_Gs_at_ipo = [1,2,3,4,5,6,7,8]
E_Gs_at_ipo = [(1,2),(1,3),(1,4),(2,6),(2,7),(3,6),(3,8),(4,7),(4,8),(5,6),(5,7),(5,8)]

V_GsF__dn_ = [1,2,3,4,5,6,7]
E_GsF__dn_ = [(1,2),(1,5),(1,7),(2,6),(2,7),(3,4),(3,6),(3,7),(4,5),(4,7),(5,6)]

cornercases = [[V_Empty, E_Empty], [V_OneNode, E_OneNode], [V_Edgeless, E_Edgeless], [V_Clique,  E_Clique]]


#random graphs
def randomGNP(n, p):
    import random
    V = []
    E = []
    for i in range(0, n):
        V.append(i)
        for j in range(i+1,n):
            if random.randint(0, 100) <= p*100:
                E.append((i, j))
    return V, E
