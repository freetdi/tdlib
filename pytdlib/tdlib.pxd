from libcpp.vector cimport vector
from libcpp cimport bool

cdef extern from "python_tdlib.hpp":

##############################################################
############ PREPROCESSING ###################################

    int gc_preprocessing(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                         vector[vector[int]] &bags, int lb, unsigned graphtype)
    int gc_PP_MD(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                 vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)
    int gc_PP_FI(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                 vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)
    int gc_PP_FI_TM(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                    vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)

##############################################################
############ LOWER BOUNDS ####################################

    int gc_deltaC_min_d(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)
    int gc_deltaC_max_d(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)
    int gc_deltaC_least_c(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)

    int gc_LBN_deltaC(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)
    int gc_LBNC_deltaC(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)
    int gc_LBP_deltaC(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)
    int gc_LBPC_deltaC(vector[unsigned int] &V, vector[unsigned int] &E, unsigned graphtype)

##############################################################
############ EXACT ALGORITHMS ################################

    int gc_exact_decomposition_cutset(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                      vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)
    int gc_exact_decomposition_cutset_decision(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                      vector[vector[int]] &V_T, vector[unsigned int] &E_T, int k, unsigned graphtype)
    int gc_exact_decomposition_dynamic(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                      vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)
    #int gc_exact_decomposition_branch_and_bound(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
    #                                            vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)

##############################################################
############ APPROXIMATIVE ALGORITHMS ########################

    int gc_seperator_algorithm(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)
    void gc_minDegree_decomp(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                               vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)
    void gc_boost_minDegree_decomp(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                               vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)
    void gc_fillIn_decomp(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                            vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)
    void gc_minDegree_ordering(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                               vector[unsigned int] &elim_ordering, unsigned graphtype)
    void gc_fillIn_ordering(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                            vector[unsigned int] &elim_ordering, unsigned graphtype)
    #int gc_random_branch_and_bound(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
    #                        vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)
    #int gc_random_elimination_orderings(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
    #                                    vector[vector[int]] &V_T, vector[unsigned int] &E_T, int lb, unsigned graphtype)

##############################################################
############ POSTPROCESSING ##################################

    int gc_MSVS(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)

    void gc_minimalChordal(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                           vector[unsigned int] &old_elimination_ordering,
                           vector[unsigned int] &new_elimination_ordering, unsigned graphtype)

##############################################################
############ APPLICATIONS ####################################

    void gc_max_clique_with_treedecomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                              vector[vector[int]] &V_T, vector[unsigned int] &E_T,
                                              vector[unsigned int] &C, unsigned graphtype)
    void gc_max_independent_set_with_treedecomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                                       vector[vector[int]] &V_T, vector[unsigned int] &E_T,
                                                       vector[unsigned int] &IS, unsigned graphtype)
    void gc_min_vertex_cover_with_treedecomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                                    vector[vector[int]] &V_T, vector[unsigned int] &E_T,
                                                    vector[unsigned int] &VC, unsigned graphtype)
    void gc_min_coloring_with_treedecomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                                    vector[vector[int]] &V_T, vector[unsigned int] &E_T,
                                                    vector[vector[int]] &col, unsigned graphtype)
    void gc_min_dominating_set_with_treedecomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                                       vector[vector[int]] &V_T, vector[unsigned int] &E_T,
                                                       vector[unsigned int] &DS, unsigned graphtype)

##############################################################
############ MISC ############################################

    int gc_ordering_to_treedec(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                               vector[vector[int]] &V_T, vector[unsigned int] &E_T,
                               vector[unsigned int] &elim_ordering, unsigned graphtype)
    void gc_treedec_to_ordering(vector[vector[int]] &V, vector[unsigned int] &E,
                                vector[unsigned int] &elim_ordering);
    int gc_validate_treedecomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                      vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)
    void gc_trivial_decomposition(vector[unsigned int] &V_G, vector[unsigned int] &E_G,
                                  vector[vector[int]] &V_T, vector[unsigned int] &E_T, unsigned graphtype)
    int gc_get_width(vector[vector[int]] &V_T)


