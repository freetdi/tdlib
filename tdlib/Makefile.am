include $(top_srcdir)/cython.am

SUBDIRS = . tests

instdir = $(pythondir)/tdlib
inst_PYTHON = \
	__init__.py \
	graphs.py \
	Decomp.py \
	Graph.py

inst_LTLIBRARIES = cytdlib.la
cytdlib_la_SOURCES = cytdlib.pyx python_tdlib.cpp

PACKAGE_STRING+= python interface

CLEANFILES += cytdlib.cpp

clean-local:
	rm -rf __pycache__

.PHONY: clean-extra symlinks

AM_CXXFLAGS = -pthread -Wl,-O1 -Wl,-Bsymbolic-functions \
              -fno-strict-aliasing -g -fwrapv -O2 -Wall \
              -Wformat -Werror=format-security

longcheck: all-am
	$(MAKE) -C tests $@

if VPATH_BUILD
SYMLINKS = symlinks
endif
all-local: $(SYMLINKS)

symlinks:
	ln -sf $(srcdir)/graphs.py
	ln -sf $(srcdir)/Graph.py
	ln -sf $(srcdir)/Decomp.py
	ln -sf $(srcdir)/printer.py
	ln -sf $(srcdir)/__init__.py

AM_CPPFLAGS = @PYTHON_INCLUDES@ -I$(top_srcdir)/src
AM_LDFLAGS = -module -omit-version

EXTRA_DIST = \
	graphs.py \
	Graph.py \
	Decomp.py \
	python_tdlib.hpp \
	cytdlib.pxd \
	cytdlib.pyx \
	printer.py

uninstall-hook:
	-rmdir ${DESTDIR}/${instdir}
	-rmdir ${DESTDIR}/${pythondir}
