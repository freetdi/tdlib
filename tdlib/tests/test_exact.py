import base
import sys
import tdlib
import unittest

from graphs import *

#don't confuse python unittest
sys.argv=sys.argv[:1]

class TestTdLib_exact(unittest.TestCase):
    def test_exact_decomposition_cutset_0(self):
        i = 0
        for V, E in cornercases:
            G = Graph(V, E)
            T, w = tdlib.exact_decomposition_cutset(G)
            self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
            print("corner case", i, w)
            i+=1

    def test_exact_decomposition_cutset_1(self):
        G = Graph(V_P6, E_P6)
        T, w = tdlib.exact_decomposition_cutset(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        print("P6", w)
        self.assertEqual(w, 1)

    def test_exact_decomposition_cutset_3(self):
        G = Graph(V_Petersen, E_Petersen)
        T, w = tdlib.exact_decomposition_cutset(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        print("Petersen", w)
        self.assertEqual(w, 4)

    def test_exact_decomposition_cutset_4(self):
        G = Graph(V_Petersen_double, E_Petersen_double)
        T, w = tdlib.exact_decomposition_cutset(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        print("Petersen_double", w)
        self.assertEqual(w, 4)

    def test_exact_decomposition_cutset_5(self):
        G = Graph(V_Wagner, E_Wagner)
        T, w = tdlib.exact_decomposition_cutset(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        print("Wagner", w)
        self.assertEqual(w, 4)

    """ takes too long
    def test_exact_decomposition_cutset_6(self):
        G = Graph(V_Pappus, E_Pappus)
        T, w = tdlib.exact_decomposition_cutset(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        print("Pappus", w)
        self.assertEqual(w, 6)
    """

    def test_exact_decomposition_cutset_7(self):
        G = Graph(V_Grid_5_5, E_Grid_5_5)
        T, w = tdlib.exact_decomposition_cutset(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        print("Grid_5_5", w)
        self.assertEqual(w, 5)

    def test_exact_decomposition_cutset_8(self):
        for n in range(0, 13):
            for i in range(0, 10):
                V, E = randomGNP(n, 0.2)
                G = Graph(V, E)
                T, w = tdlib.exact_decomposition_cutset(G)
                self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
            print("random..")

    def test_exact_decomposition_cutset_9(self):
        status = True
        for n in range(0, 13):
            for i in range(0, 10):
                V, E = randomGNP(n, 0.3)
                G = Graph(V, E)
                T, w2 = tdlib.PP_FI_TM(G)
                isleq = tdlib.exact_decomposition_cutset_decision(G, w2)
                if(not isleq):
                    print("error [validate width], graph: " + str(G.vertices()) + ", " + str(G.edges()))
                    print("width_PP_FI_TM: " + str(w2))
                    N, M, width = tdlib.exact_decomposition_cutset(G)
                    hrgl = tdlib.check_treedec(V, E, N, M, message=True)
                    print("proper width_PP_FI_TM: " + str(width) + " " + str(hrgl))
                    self.assertEqual(hrgl, 0)
                    status = False
            print("random..")

        self.assertEqual(status, True)

    def test_exact_decomposition_dynamic_0(self):
        print("=== dynamic")
        i = 0
        for V, E in cornercases:
            G = Graph(V, E)
            T, w = tdlib.exact_decomposition_dynamic(G)
            self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
            print("corner", i, w)
            i+=1

        G = Graph(V_P6, E_P6)
        T, tw = tdlib.exact_decomposition_dynamic(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        self.assertEqual(tw, 1)
        print("P6", w)

        G = Graph(V_K5, E_K5)
        T, tw = tdlib.exact_decomposition_dynamic(G)
        self.assertEqual(tdlib.is_valid_treedecomposition(G, T), True)
        self.assertEqual(tw, 4)

        print("K5", w)

if __name__ == '__main__':
    unittest.main()

# vim:ts=4:sw=4:et
