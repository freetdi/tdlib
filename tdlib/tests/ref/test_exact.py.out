corner case 0 -1
corner case 1 0
corner case 2 0
corner case 3 4
P6 1
Petersen 4
Petersen_double 4
Wagner 4
Grid_5_5 5
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
random..
=== dynamic
corner 0 -1
corner 1 0
corner 2 0
corner 3 4
P6 4
K5 4
